src/bc_protocol/time.rs

```diff
@@ -0,0 +1,132 @@
+use super::{BcCamera, Error, Result, RX_TIMEOUT};
+use crate::bc::{model::*, xml::*};
+use time::{date, Date, OffsetDateTime, PrimitiveDateTime, Time, UtcOffset};
+
+impl BcCamera {
+    pub fn get_time(&self) -> Result<Option<OffsetDateTime>> {
+        let connection = self
+            .connection
+            .as_ref()
+            .expect("Must be connected to get time");
+        let sub_get_general = connection.subscribe(MSG_ID_GET_GENERAL)?;
+        let get = Bc {
+            meta: BcMeta {
+                msg_id: MSG_ID_GET_GENERAL,
+                client_idx: 0,
+                encrypted: true,
+                class: 0x6414,
+            },
+            body: BcBody::ModernMsg(ModernMsg::default()),
+        };
+
+        sub_get_general.send(get)?;
+        let msg = sub_get_general.rx.recv_timeout(RX_TIMEOUT)?;
+
+        if let BcBody::ModernMsg(ModernMsg {
+            xml:
+                Some(BcXml {
+                    system_general:
+                        Some(SystemGeneral {
+                            time_zone: Some(time_zone),
+                            year: Some(year),
+                            month: Some(month),
+                            day: Some(day),
+                            hour: Some(hour),
+                            minute: Some(minute),
+                            second: Some(second),
+                            ..
+                        }),
+                    ..
+                }),
+            ..
+        }) = msg.body
+        {
+            let datetime = match try_build_timestamp(
+                time_zone, year, month, day, hour, minute, second
+            ) {
+                Ok(dt) => dt,
+                Err(e) => return Err(Error::UnintelligibleReply {
+                    reply: msg,
+                    why: "Could not parse date",
+                })
+            };
+
+            // This code was written in 2020; I'm trying to catch all the possible epochs that
+            // cameras might reset themselves to. My B800 resets to Jan 1, 1999, but I can't
+            // guarantee that Reolink won't pick some newer date.  Therefore, last year ought
+            // to be new enough, yet still distant enough that it won't interfere with anything
+            const BOUNDARY: Date = date!(2019-01-01);
+
+            // detect if no time is actually set, and return Ok(None): that is, operation
+            // succeeded, and there is no time set
+            if datetime.date() < BOUNDARY {
+                Ok(None)
+            } else {
+                Ok(Some(datetime))
+            }
+        } else {
+            Err(Error::UnintelligibleReply {
+                reply: msg,
+                why: "Reply did not contain SystemGeneral with all time fields filled out",
+            })
+        }
+    }
+
+    pub fn set_time(&self, timestamp: OffsetDateTime) -> Result<()> {
```

Comment:
Will setting time fail if we don't have admin priv?

Reply:
Oh that's a good question. I have not tried that.

Reply:
Well, it doesn't really matter what's going to happen here if we don't have permission. We do recheck that the action "took" and print an error if not. So I think that's okay.

Reply:
I see, then am wondering if it would help the user for that message to give a potential reason for not being able setting it. Else we may have issue reports asking for help with their error.

Reply:
Good point. Done.